File: base/src/main/java/com/santiihoyos/base/di/ViewModelKey.java
package com.santiihoyos.base.di;

import com.santiihoyos.base.abstracts.BaseViewModel;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import dagger.MapKey;

@MapKey
@Documented
@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
public @interface ViewModelKey {
    Class<? extends BaseViewModel> value();
}

File: base/src/main/java/com/santiihoyos/base/di/ViewModelKey.kt
package com.santiihoyos.base.di

import com.santiihoyos.base.feature.abstracts.BaseViewModel
import dagger.MapKey
import kotlin.reflect.KClass

@MapKey
@Target(
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER
)
@Retention(AnnotationRetention.RUNTIME)
annotation class ViewModelKey(val value: KClass<out BaseViewModel>)
